### PR TITLE
refactor: let verifiers know the real tip number and tx status

### PR DESCRIPTION
**Pull request by yangby-cryptape**
### Purpose

In the latest version, we hard-code `tip_number + 1` as the block number for verification.

For example:
https://github.com/nervosnetwork/ckb/blob/dd9040f207364a039fdbe5fa9a21510c26b67b8f/tx-pool/src/util.rs#L86-L93
https://github.com/nervosnetwork/ckb/blob/dd9040f207364a039fdbe5fa9a21510c26b67b8f/tx-pool/src/util.rs#L99-L106

There are several defects:

- The tip header hash doesn't correspond to the tip block number.

- The epoch number doesn't correspond to the tip block number.

- For hard fork at a particular epoch, the script verifier have to know the exact epoch number.

  But epoch number couldn't add a number directly, since a epoch length is depend on the whole previous epoch information.

  So, the best way, I think, is to pass the real epoch number to all verifiers and tell those verifiers the minimum block latency.

  Then, I think the block number should be checked with same logic.

### Further

The `MaturityVerifier` and `SinceVerifier` with epoch numbers could be optimized, too.

### CODE REVIEW

File: verification/src/transaction_verifier.rs [hunk 1]
```diff
@@ -20,6 +20,59 @@ use lru::LruCache;
 use std::cell::RefCell;
 use std::collections::HashSet;
 
+/// The phase that transactions are in.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum TransactionVerificationPhase {
+    /// The transaction has just been submitted.
+    ///
+    /// So the transaction will be:
```
Comment (keroro520) [inline verification/src/transaction_verifier.rs:28]: unless traffic jam?

Comment (keroro520) [reply verification/src/transaction_verifier.rs:28]: Confused with `Fresh`, `Submitted`, `Pending` ...

File: verification/src/transaction_verifier.rs [hunk 2]
```diff
@@ -640,7 +702,7 @@ impl<'a, DL: HeaderProvider> SinceVerifier<'a, DL> {
         if since.is_absolute() {
             match since.extract_metric() {
                 Some(SinceMetric::BlockNumber(block_number)) => {
-                    if self.block_number < block_number {
+                    if self.tx_phase.earliest_committed_block(self.block_number) < block_number {
```
Comment (keroro520) [inline verification/src/transaction_verifier.rs:705]: Pay attention to this. You change the `block_number` from `tip_number + 1` to `tip_number` [outside](https://github.com/nervosnetwork/ckb/pull/2652/files#diff-f17291b75b5bafb128afde15b639dea00e71c718713a389060fe31cd9e53a914R103).

As a result, after the above change, block_number is 1 less than the old code.
I think here it should use `<=`.

**Review by keroro520 — CHANGES_REQUESTED** [review]

File: verification/src/transaction_verifier.rs [hunk 3]
```diff
@@ -20,6 +20,59 @@ use lru::LruCache;
...
+    /// So the transaction will be committed after the `tip_number + 1` block.
+    Proposed,
+    /// The transaction is commit.
+    ///
+    /// So the transaction will be committed in current block.
+    Committed,
+}
+
+impl TransactionVerificationPhase {
+    /// Convert tip block number to earliest committed block number.
```
[the remark this answers is not in the document]
Comment (yangby-cryptape) [reply verification/src/transaction_verifier.rs:47]: The explanation of incremental blocks is already in the enumerated type definition. I think there are no need to make a copy.

File: verification/src/transaction_verifier.rs (hunk 2 above)
Comment (yangby-cryptape) [reply verification/src/transaction_verifier.rs:705]: I think the integration test  `ValidSince` could prove that current comparison is correct.

File: test/src/specs/tx_pool/valid_since.rs [hunk 4]
```diff
@@ -56,11 +60,29 @@ impl ValidSince {
                 .is_ok(),
             "transaction is ok, tip is equal to relative since block number",
         );
+
+        // Pending
+        node.assert_tx_pool_size(1, 0);
+        assert!(!is_transaction_committed(node, &transaction));
```
[the remark this answers is not in the document]
Comment (yangby-cryptape) [reply test/src/specs/tx_pool/valid_since.rs]: DONE!

**Review by keroro520 — APPROVED** [review]

**Comment by nervos-bot-user** [comment]
Can one of the admins verify this patch?

**Review by driftluo — DISMISSED** [review]

File: verification/src/transaction_verifier.rs [hunk 5]
```diff
@@ -20,6 +20,59 @@ use lru::LruCache;
...
+    }
+
+    /// Convert tip epoch number to earliest committed epoch number.
+    pub fn earliest_committed_epoch(self, tip_epoch: EpochNumberWithFraction) -> EpochNumber {
+        let number = tip_epoch.number();
+        let length = tip_epoch.length();
+        let index = tip_epoch.index();
+        let index_added = match self {
+            Self::Submitted => 3,
+            Self::Gap => 2,
```
Comment (doitian) [inline verification/src/transaction_verifier.rs:64]: The gap is not a constant but is configurable in the consensus via the option `tx_proposal_window`

File: verification/src/transaction_verifier.rs [hunk 6]
```diff
@@ -20,6 +20,59 @@ use lru::LruCache;
...
+        }
+    }
+
+    /// Convert tip epoch number to earliest committed epoch number.
+    pub fn earliest_committed_epoch(self, tip_epoch: EpochNumberWithFraction) -> EpochNumber {
+        let number = tip_epoch.number();
+        let length = tip_epoch.length();
+        let index = tip_epoch.index();
+        let index_added = match self {
+            Self::Submitted => 3,
```
Comment (doitian) [inline verification/src/transaction_verifier.rs:63]: This is not a constant but is configurable in the consensus. It should be set to `tx_proposal_window.closest + 1`

File: verification/src/transaction_verifier.rs [hunk 7]
```diff
@@ -20,6 +20,59 @@ use lru::LruCache;
...
+    /// So the transaction will be committed in current block.
+    Committed,
+}
+
+impl TransactionVerificationPhase {
+    /// Convert tip block number to earliest committed block number.
+    pub fn earliest_committed_block(self, tip_block_number: BlockNumber) -> BlockNumber {
+        match self {
+            Self::Submitted => tip_block_number + 3,
+            Self::Gap => tip_block_number + 2,
```
Comment (doitian) [inline verification/src/transaction_verifier.rs:51]: The tx proposal window is configurable in the consensus, see comment below.

File: verification/src/transaction_verifier.rs [hunk 8]
```diff
@@ -20,6 +20,59 @@ use lru::LruCache;
...
+    ///
+    /// So the transaction will be committed in current block.
+    Committed,
+}
+
+impl TransactionVerificationPhase {
+    /// Convert tip block number to earliest committed block number.
+    pub fn earliest_committed_block(self, tip_block_number: BlockNumber) -> BlockNumber {
+        match self {
+            Self::Submitted => tip_block_number + 3,
```
Comment (doitian) [inline verification/src/transaction_verifier.rs:50]: The tx proposal window is configurable in the consensus, see comment below.

**Review by doitian — CHANGES_REQUESTED** [review]

File: verification/src/transaction_verifier.rs (hunk 6 above)
Comment (yangby-cryptape) [reply verification/src/transaction_verifier.rs]: DONE!

File: verification/src/transaction_verifier.rs [hunk 9]
```diff
@@ -589,29 +659,36 @@ impl Since {
...
     block_number: BlockNumber,
     epoch_number_with_fraction: EpochNumberWithFraction,
     parent_hash: Byte32,
+    tx_phase: TransactionVerificationPhase,
     median_block_count: usize,
     median_timestamps_cache: RefCell<LruCache<Byte32, u64>>,
 }
 
 impl<'a, DL: HeaderProvider> SinceVerifier<'a, DL> {
+    #[allow(clippy::too_many_arguments)]
```
Comment (yangby-cryptape) [inline verification/src/transaction_verifier.rs:672]: There are 4 verifiers have `#[allow(clippy::too_many_arguments)]`. All of them have 4 same arguments, three of them have 7 same arguments.
Could do refactor later, in another PR :dog:.

**Review by driftluo — APPROVED** [review]

**Comment by yangby-cryptape** [comment]
We can NOT get the parent hash for a transaction which will committed after 2 blocks.

So this PR is not right.

**Pull request closed without merging**